### ISSUE TITLE
A Primer in BERTology: What we know about how BERT works

### ISSUE THREAD

**Issue by amyxie361** [amyxie361/Blog#17]
[A Primer in BERTology: What we know about how BERT works](https://arxiv.org/pdf/2002.12327.pdf)

关于BERT可解释性的survey

3. BERT embedding

 - Gregor Wiedemann, (2019): 
BERT’s contextualized embeddings form distinct and clear clusters corresponding to **word senses**?
- Mickus et al. (2019):
representation of the same word varies depending on position of the sentence in
which it occurs
- **Ethayarajh** (2019):
how similar the embeddings for identical words are in every layer: produce more context-specific representations, embeddings occupy a narrow cone ???

4.1 Syntactic knowledge

- Lin et al. (2019) : 
BERT representations are hierarchical rather than linear: akin to syntactic tree. (Does that meen we don't need trees any more?)
- Tenney et al. (2019b) and Liu et al. (2019a):
BERT embeddings encode information about parts of speech, syntactic chunks
and roles, but BERT’s knowledge of syntax is partial. (cannot cover distant parent node)
- the syntactic structure is not directly encoded in self-attention weights, but they can be transformed to reflect it.
- Htut et al. (2019)
unable to extract full parse trees from BERT heads even with the gold annotations for the root
- Jawahar et al. (2019)
 dependency tree extracted directly from self-attention weights, but provide no quantitative evaluation
- **Hewitt and Manning** (2019)
learn transformation matrices that would successfully recover much of the Stanford Dependencies formalism for PennTreebank data( 为什么这里又可以了)
- Jawahar et al. (2019)
approximate BERT representations with Tensor Product Decomposition Networks:
dependency trees are the best match among 5 decomposition schemes
- Goldberg (2019) 
BERT takes subject-predicate agreement into account when performing the cloze task
- Warstadt et al. (2019)
BERT is better able to detect the presence of NPIs (e.g. ”ever”) and the words that allow their use (e.g. ”whether”) than scope violations. ???
- (**Ettinger,** 2019) 
predictions were not altered even with shuffled word order, truncated sentences, removed subjects and objects (矛盾？bert到底懂不懂结构)
- (Wallace et al., 2019a)
adversarial attacks, with models disturbed by nonsensical inputs

4.2 Semantic knowledge

- (Ettinger, 2019)
BERT has some knowledge for semantic roles
prefer the incorrect fillers for semantic roles that are semantically related to the correct
ones, to those that are unrelated (e.g. “to tip a chef” should be better than “to tip a robin”, but worse than “to tip a waiter”). ???
- Tenney et al. (2019b)
BERT encodes information about entity types, relations, semantic roles, and proto-roles : can be classified.
- (Wallace et al., 2019b)
BERT struggles with representations of numbers. 
Addition and number decoding tasks,
does not form good representations for floating point numbers and fails to generalize away from the training data(一部分来自于wordpiece)

4.3 world knowledge
- Petroni et al. (2019) 
for some relation types, vanilla BERT is competitive with methods relying on knowledge bases(relation type 到底是relation还是knowledge)
- (Bouraoui et al., 2019; Jiang et al.) automatic extraction and augmentation 
 need good template sentences to extract knowledge.
- BERT cannot reason based on its world knowledge.
- Forbes et al. (2019) 
can “guess” the affordances and properties of many objects, 
but does not have the information about their interactions
 (e.g. it “knows” that people can walk into houses, and that houses are big, but it cannot infer that houses are bigger than people.)
can't do inference, performance drops as more inference steps.
- Poerner et al. (2019):
 success in factoid knowledge retrieval comes from learning **stereotypical character combinations**, e.g. it would predict that a person with an Italian-sounding name is Italian, even when it is factually incorrect.

5 Localizing linguistic knowledge

5.1 self-attention heads

 classification of attention head types:
- (Raganato and Tiedemann, 2018): attending to the word itself, to previous/next words and to the end of the sentence
- (Clark et al., 2019): attending to previous/next tokens, [CLS], [SEP], punctuation, and “attending broadly” over the sequence 
- **(Kovaleva et al., 2019)** Vertical(much- to cls/sep/punctuation), Diagonal, Vertical+Diagonal, Block, Heterogeneous(less than half).
- Kovaleva et al. (2019) 
most self-attention heads do not directly encode any nontrivial linguistic information,
(overparameterization)
- **Clark** et al. (2019) 
the function of [SEP] might be one of “no-op”, a signal to ignore the head if its pattern is not applicable to the current case
- Some BERT heads seem to specialize in certain types of syntactic relations.
 attended significantly more than a random baseline to words in certain syntactic positions
(attend to words in **obj** more ) (positional baseline 是啥)
no single head has the complete syntactic tree information
- Lin et al. (2019) 
attention weights are weak indicators of subject-verb agreement and reflexive anaphora.
- Clark et al. (2019)
 identify a BERT head that can be directly used as a classifier to perform coreference resolution on par with a rule-based system
- **Kovaleva** et al. (2019)
even when attention heads specialize in tracking semantic relations, they do not necessarily contribute to BERT’s performance on relevant tasks.？？？

5.2 BERT layers

- lower layers have the most linear word order information
Lin et al. (2019) a decrease in the knowledge of linear word order around layer 4 in BERT-base.
predicting the index of a token, the main auxiliary verb and the sentence subject
- syntactic information is the most prominent in the middle BERT layers. 
Hewitt and Manning (2019) -  syntactic tree - base(6-9), large(14-19)
Goldberg (2019) - subject-verb agreement - 8-9
Jawahar et al. (2019) - syntactic probing tasks - middle
- **Liu et al. (2019a)** 
middle layers of Transformers are overall the best-performing and the most transferable across tasks(6-9?)
![image](https://user-images.githubusercontent.com/26669286/75570125-7ff05200-5a24-11ea-86e2-41fa39d7da2d.png)
- **Tenney et al. (2019a)**
the basic syntactic information appears earlier in the network while high-level semantic features appears at the higher layers
- Jawahar et al. (2019) 
lower layers were more useful for chunking, 
while middle layers were more useful for parsing
- Liu et al. (2019a) find the opposite: both POS tagging and chunking were also performed best at the middle layers, in both BERT-base and BERTlarge. ???
- The final layers of BERT are the most task specific
(为什么在pretrain的时候前面的层没学MLM？)
- Hao et al. (2019)  weights of lower layers of the fine-tuned BERT are restored to their original values, it does not dramatically hurt the model performance.

**Comment by amyxie361** [comment]
[中文翻译总结](https://thinkwee.top/2020/03/02/bertology/#more)
